src/ray/util/event.cc

```diff
@@ -0,0 +1,106 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/util/event.h"
+
+namespace ray {
+
+EventManager &EventManager::Instance() {
+  static EventManager instance_;
+  return instance_;
+}
+
+bool EventManager::IsEmpty() { return reporter_map_.empty(); }
+
+void EventManager::Publish(rpc::Event &event) {
+  for (const auto &element : reporter_map_) {
+    (element.second)->Report(event);
+  }
+}
+
+void EventManager::AddReporter(std::shared_ptr<LogBasedEventReporter> reporter) {
+  reporter_map_.emplace(reporter->GetReporterKey(), reporter);
+}
+
+void EventManager::ClearReporters() { reporter_map_.clear(); }
+
+thread_local std::unique_ptr<RayEventContext> RayEventContext::context_ = nullptr;
+
+RayEventContext &RayEventContext::Instance() {
+  if (context_ == nullptr) {
+    context_ = std::unique_ptr<RayEventContext>(new RayEventContext());
+  }
+  return *context_;
+}
+
+void RayEventContext::SetEventContext(
+    rpc::Event_SourceType source_type,
+    std::unordered_map<std::string, std::string> custom_field) {
+  source_type_ = source_type;
+  custom_field_ = custom_field;
+}
+
+void RayEventContext::ResetEventContext() {
+  source_type_ = rpc::Event_SourceType::Event_SourceType_COMMON;
+  custom_field_.clear();
+}
+
+void RayEventContext::SetCustomField(std::string key, std::string value) {
+  custom_field_[key] = value;
+}
+
+void RayEventContext::SetCustomField(
+    std::unordered_map<std::string, std::string> custom_field) {
+  custom_field_ = custom_field;
+}
+
+void RayEvent::ReportEvent(std::string severity, std::string label, std::string message) {
+  rpc::Event_Severity severity_ele =
+      rpc::Event_Severity::Event_Severity_Event_Severity_INT_MIN_SENTINEL_DO_NOT_USE_;
+  RAY_CHECK(rpc::Event_Severity_Parse(severity, &severity_ele));
+  RayEvent(severity_ele, label) << message;
+}
+
+RayEvent::~RayEvent() { SendMessage(osstream_.str()); }
+
+void RayEvent::SendMessage(const std::string &message) {
+  RAY_CHECK(rpc::Event_SourceType_IsValid(RayEventContext::Instance().GetSourceType()));
+  RAY_CHECK(rpc::Event_Severity_IsValid(severity_));
+
+  if (EventManager::Instance().IsEmpty()) {
+    return;
+  }
+
+  rpc::Event event;
+
+  std::string event_id_buffer = std::string(18, ' ');
+  FillRandom(&event_id_buffer);
+  event.set_event_id(StringToHex(event_id_buffer));
+
+  event.set_source_type(RayEventContext::Instance().GetSourceType());
+  event.set_source_hostname(RayEventContext::Instance().GetSourceHostname());
+  event.set_source_pid(RayEventContext::Instance().GetSourcePid());
+
+  event.set_severity(severity_);
+  event.set_label(label_);
+  event.set_message(message);
+  event.set_timestamp(current_sys_time_us());
+
+  auto mp = RayEventContext::Instance().GetCustomField();
+  event.mutable_custom_field()->insert(mp.begin(), mp.end());
+
+  EventManager::Instance().Publish(event);
```

Comment:
This seems to have pretty big impact when we have it on the critical path. I don't think we should care about it in this PR, but we should consider optimization later. Thoughts? @ashione @edoakes

Reply:
In Ant, we started to consider using io_service and open a separate thread to handle the sending logic.

Reply:
That sounds good to me! Maybe we can share the current stats io service for this.

Reply:
OK, I prefer to discuss this in the next PR. I think this will be a bit complicated, mainly about how to make buffers and send events regularly.